File: src/bin/0070-climbing-stairs.py
# @leetup=info id=70 lang=python3 slug=climbing-stairs



# @leetup=code
class Solution:
    def climbStairs(self, n: int) -> int:
        # dp = [0] * (n + 1)
        # dp[0] = 1
        # dp[1] = 1
        # for i in range(2, n + 1):
        #     dp[i] = dp[i - 1] + dp[i - 2]
        # return dp[n]

        dp0 = 1
        dp1 = 1
        for _ in range(2, n + 1):
            (dp1, dp0) = (dp0 + dp1, dp1)
        return dp1


# @leetup=code

File: src/bin/0070-climbing-stairs.rs
// @leetup=info id=70 lang=rust slug=climbing-stairs

// @leetup=code
impl Solution {
    pub fn climb_stairs(n: i32) -> i32 {
        let mut dp0 = 1;
        let mut dp1 = 1;
        for _ in 2..n + 1 {
            // [2023-04-13 Thu] leetcode's rust doesn't support this:
            // destructuring assignments are unstable (solution.rs)
            // (dp1, dp0) = (dp0 + dp1, dp1);

            let tmp = dp0 + dp1;
            dp0 = dp1;
            dp1 = tmp;
        }
        dp1
    }
}
// @leetup=code

struct Solution;

fn main() {
    let _ = Solution::climb_stairs;
}
